tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -1041,13 +1046,25 @@ unsigned CGMSHLSLRuntime::ConstructStructAnnotation(DxilStructAnnotation *annota
       }
     }
 
-    unsigned arrayEltSize = 0;
     // Process field to make sure the size of field is ready.
+    unsigned arrayEltSize = 0;
     unsigned size = AddTypeAnnotation(fieldDecl->getType(), dxilTypeSys, arrayEltSize);
 
-    // Update offset.
-    offset += size;
-    
+    // Align offset.
+    if (size) {
+      unsigned offset = AlignBaseOffset(fieldTy, CBufferOffset, bDefaultRowMajor, CGM, dataLayout);
+      if (packOffset && CBufferOffset != offset) {
+        // custom offset has an alignment problem, or this code does
```

Comment:
This is meant to catch a potential error in user packoffset, but if that code is never hit here, then this code isn't either, so after verifying whether that is the case, we could end up deleting the packoffset path entirely from here.  When I tried to test for this case, it didn't hit the packoffset code here, and it crashed elsewhere.  I didn't have a good localized fix, so I didn't end up fixing that issue in this PR.  I left this in in case there was an equivalent for struct fields I wasn't hitting, or in case I could reuse this code for that case in the future.